Opaque types with resilience

jckarter opened the pull request:
A continuation of #21137. This branch adds "proper" resilience support. Opaque types can now enter the SIL type system, and will be able to be represented at runtime with opaque accessors for their associated metadata.

jckarter committed:
Parse opaque types.

jckarter committed:
AST: Represent OpaqueTypeDecls.

To represent the abstracted interface of an opaque type, we need a generic signature that refines
the outer context generic signature with an additional generic parameter representing the underlying
type and its exposed constraints. Opaque types also need to be keyed by their originating decl, so
that we can treat values of the same opaque type as the same. When we check a FuncDecl with an
opaque type specified as its return type, create an OpaqueTypeDecl and associate it with the
originating decl. (A representation for *types* derived from the opaque decl will come next.)

jckarter committed:
Add an OpaqueTypeArchetypeType subclass.

jckarter committed:
Sema: Infer the underlying type for opaque return types from function bodies.

jckarter committed:
SILGen: Substitute away opaque types.

jckarter committed:
Support nested types on opaque archetypes (and maybe opened ones).

jckarter committed:
Serialization: Serialize opaque types and their decls.

jckarter committed:
IRGen: Substitute out opaque types when emitting associated types.

jckarter committed:
Enable opaque result types.

jckarter committed:
Respond to Slava's feedback

jckarter committed:
Mangle opaque types without underlying types as ErrorType for now.

jckarter committed:
Remove ResilienceExpansion from substOpaqueTypes for now.

It's currently meaningless, and it'll require thought to pass the correct value when it becomes
meaningful.

jckarter committed:
Sema: Don't allow properties with inferred types to infer opaque result types.

This prevents opaque result types from propagating nontrivially into other declarations' types,
which may be confusing and create implementation complexities.

jckarter committed:
Add test for deserializing opaque types across modules

jckarter committed:
SILGen: Start supporting opaque result types resiliently.

Tear out the hacks to pre-substitute opaque types before they enter the SIL type system.
Implement UnderlyingToOpaqueExpr as bitcasting the result of the underlying expression from the
underlying type to the opaque type.

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -170,6 +171,113 @@ static void revertDependentTypeLoc(TypeLoc &tl) {
+    }
+  }
+  
+  auto interfaceSignature = GenericSignature::get(interfaceGenericParams,
```
It might be cleaner if you use a GSB instead; then you can add a requirement with a class or any kind of existential in it and it will flatten it into the right thing. In particular, there's no guarantee that the order in which you're adding stuff above is the right order to get a canonical generic signature

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -170,6 +171,113 @@ static void revertDependentTypeLoc(TypeLoc &tl) {
+  }
+  auto opaqueTy = OpaqueTypeArchetypeType::get(opaqueDecl, subs);
+  auto metatype = MetatypeType::get(opaqueTy);
+  opaqueDecl->setInterfaceType(metatype);
```
The code for setting the type here is the same as in deserialization. Take a look at how NominalTypeDecl and AssociatedTypeDecl have `computeType()` methods; OpaqueTypeDecl should also have one.

slavapestov commented on lib/Sema/TypeCheckProtocolInference.cpp:
```diff
@@ -1966,8 +1966,9 @@ auto AssociatedTypeInference::solve(ConformanceChecker &checker)
         return None;
 
-      if (replacement->hasArchetype())
+      if (replacement->hasArchetype()) {
```
Useless edit here

jckarter replied:
I thought LLVM style was always to brace if blocks. It's more maintainable regardless.

slavapestov commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -3344,12 +3346,51 @@ ModuleFile::getDeclCheckedImpl(DeclID DID) {
+                     getGenericSignature(interfaceSigID),
+                     getType(interfaceTypeID)->castTo<GenericTypeParamType>());
+    auto genericEnv = getGenericEnvironment(genericEnvID);
+    opaqueDecl->setGenericEnvironment(genericEnv);
```
For later: there's a mechanism to lazily deserialize generic environments. We do this already for generic functions and types; you might want to look into doing it here

slavapestov commented on lib/AST/ASTContext.cpp:
```diff
@@ -4283,6 +4284,139 @@ DependentMemberType *DependentMemberType::get(Type base,
+    
+    auto replacementParam = replacement->getAs<GenericTypeParamType>();
+    if (!replacementParam)
+      llvm_unreachable("opaque types cannot currently be parameterized by non-identity type parameter mappings");
```
Nitpick: lines longer than 80 characters

slavapestov commented on lib/AST/ASTContext.cpp:
```diff
@@ -4283,6 +4284,139 @@ DependentMemberType *DependentMemberType::get(Type base,
+  auto opaqueInterfaceTy = Decl->getUnderlyingInterfaceType();
+  auto layout = signature->getLayoutConstraint(opaqueInterfaceTy);
+  auto superclass = signature->getSuperclassBound(opaqueInterfaceTy);
+  SmallVector<ProtocolDecl*, 4> protos;
```
Can you just do `protos = signature->getConformsTo(opaqueInterfaceTy)`?

slavapestov commented on lib/AST/ASTContext.cpp:
```diff
@@ -4283,6 +4284,139 @@ DependentMemberType *DependentMemberType::get(Type base,
+  
+  // Create a generic environment and bind the opaque archetype to the
+  // opaque interface type from the decl's signature.
+  auto env = signature->createGenericEnvironment();
```
Possibly for later: can you lazily create the generic environment?

slavapestov commented on lib/AST/ASTContext.cpp:
```diff
@@ -4324,6 +4458,15 @@ CanOpenedArchetypeType OpenedArchetypeType::get(Type existential,
+                                               nullptr);
+  auto env = signature->createGenericEnvironment();
+  env->addMapping(signature->getGenericParams()[0], result);
+  result->Environment = env;
```
This is actually a regression. Can you lazily create the generic environment?

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -3273,9 +3295,7 @@ void TypeAliasDecl::setUnderlyingType(Type underlying) {
-    auto subs = SubstitutionMap::get(
-        genericSig, [&](SubstitutableType *type) -> Type { return type; },
-        MakeAbstractConformanceForGenericType());
+    auto subs = genericSig->getIdentitySubstitutionMap();
```
Thanks

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -5703,8 +5738,9 @@ void AbstractFunctionDecl::computeType(AnyFunctionType::ExtInfo info) {
   if (auto fn = dyn_cast<FuncDecl>(this)) {
     resultTy = fn->getBodyResultTypeLoc().getType();
-    if (!resultTy)
+    if (!resultTy) {
```
Useless edit

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -2415,6 +2430,114 @@ NestedArchetypeType::NestedArchetypeType(const ASTContext &Ctx,
+ReplaceOpaqueTypesWithUnderlyingTypes::operator()(CanType maybeOpaqueType,
+                                                 Type replacementType,
+                                                 ProtocolDecl *protocol) const {
+  auto abstractRef = ProtocolConformanceRef(protocol);
```
This is almost identical to the above

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -3531,6 +3672,55 @@ case TypeKind::Id:
+    if (!anyChanged)
+      return *this;
+    
+    // FIXME: This re-looks-up conformances instead of transforming them in
```
Perhaps subst() should handle opaque types directly and transform() should not attempt the substitution here?

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -3531,6 +3672,55 @@ case TypeKind::Id:
+      return *this;
+    }
+    
+    // Substitute the new root into the root of the interface type.
```
It feels like this should only need to happen in subst() too.

slavapestov commented on lib/SILGen/SILGenBuilder.cpp:
```diff
@@ -156,6 +155,7 @@ InitExistentialValueInst *SILGenBuilder::createInitExistentialValue(
 
 InitExistentialMetatypeInst *SILGenBuilder::createInitExistentialMetatype(
     SILLocation loc, SILValue metatype, SILType existentialType,
+    CanType formalConcreteInstanceType,
```
Do you really need this? The instance type of a metatype is unlowered, so it should match the formal instance type.

jckarter replied:
It was necessary to support the hack of substituting out opaque types everywhere in SILGen, but I can back it out now.

slavapestov commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -1916,6 +1917,66 @@ RValue RValueEmitter::visitCoerceExpr(CoerceExpr *E, SGFContext C) {
+    return RValue(SGF, E, opaque);
+  }
+  
+  // If the opaque type is loadable, emit the subexpression and bitcast it.
```
I wonder if opaque types should be more like existentials, with special instructions to project an address or wrap a loadable value...

jckarter replied:
We could add special instructions, but I don't know that it's worth it. The SIL passes already know how to deal with specializing bitcasts, and if we added specific instructions for underlying-to-opaque and the opposite conversion, they'd have to all be taught about those new instructions which are 99% semantically the same.

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2007,6 +2038,48 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
+      auto opaque1 = cast<OpaqueTypeArchetypeType>(desugar1);
+      auto opaque2 = cast<OpaqueTypeArchetypeType>(desugar2);
+      
+      assert(!type2->is<LValueType>() && "Unexpected lvalue type!");
```
type1 and type2 will never be LValueType at this point, so you can remove the cargo cult

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2007,6 +2038,48 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
+    
+    // Same for nested archetypes rooted in opaque types.
+    case TypeKind::NestedArchetype: {
+      auto nested1 = cast<NestedArchetypeType>(desugar1);
```
Do you actually check anywhere that the two nested types are the same? It looks like you're only comparing the root types, but I might be wrong

jckarter replied:
I check that the opaque decls match and that `interfaceTy1 == interfaceTy2` right below this.

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -1285,11 +1290,49 @@ ConstraintSystem::matchSuperclassTypes(Type type1, Type type2,
+  if (auto arch1 = type1->getAs<ArchetypeType>()) {
+    auto arch2 = type2->castTo<ArchetypeType>();
+    auto opaque1 = cast<OpaqueTypeArchetypeType>(arch1->getRoot());
+    auto opaque2 = cast<OpaqueTypeArchetypeType>(arch2->getRoot());
```
Do you need to handle nested types here? It looks like you use the DeepEquality restriction to match nested types as well

jckarter replied:
If we've gotten this far, then the opaque decls and nested type paths must already match, and the types at most differ in the substitutions. The assertions directly below this assert this.

jckarter committed:
Mangle opaque result types.

jckarter committed:
IRGen: Emit runtime calls to fetch opaque type metadata.

Introduce code to emit runtime calls to get the type metadata for the underlying type of an opaque type and to get its conformances.

jckarter committed:
IRGen: Use runtime calls to get opaque type metadata and conformances for associated types.

jckarter committed:
opaque descriptor wip

jckarter committed:
IRGen: Emit opaque type descriptors.

jckarter committed:
Push the conformance accessor hack down into getTypeByMangledName.

jckarter committed:
IRGen: Inline helper into its last remaining use

jckarter committed:
IRGen: Extract GenKeyPath's emitWitnessTableRefString to be usable elsewhere.

jckarter committed:
fill in witness tables

jckarter committed:
change signature

jckarter committed:
runtime interface fixup irgen

jckarter committed:
Runtime: Implement entry points for opaque type support

slavapestov commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -1655,6 +1634,99 @@ namespace {
+    }
+    
+    void addUnderlyingTypeAndConformances() {
+      auto underlyingType = Type(O->getUnderlyingInterfaceType())
```
Should this be a utility method on OpaqueTypeDecl?

jckarter committed:
Decode opaque types in the runtime demangler.

jckarter committed:
add changelog entry for opaque types

jckarter committed:
Switch __opaque to some, per the latest rev of the proposal

jckarter committed:
update tests

jckarter committed:
fix rdar://problem/49230987

jckarter committed:
IRGen: Bind local metadata before emitting opaque type assoc type witnesses. rdar://problem/49585457

jckarter committed:
TBD: Include opaque type descriptors

nkcsgexi commented on lib/Parse/ParseType.cpp:
```diff
@@ -670,19 +670,40 @@ ParserResult<TypeRepr> Parser::parseTypeIdentifier() {
+      && Tok.is(tok::identifier)
+      && Tok.getRawText() == "some") {
+    opaqueLoc = consumeToken();
+  }
```
Any reason we don't need back tracking scope for this `some`? The following code will have trouble being parsed:
```
class some {}
func foo() -> some { return some() }
```

jckarter replied:
Yes, it should have a backtracking scope.

jckarter committed:
Parsable interface and type reconstruction support for opaque types.

When printing a swiftinterface, represent opaque result types using an attribute that refers to
the mangled name of the defining decl for the opaque type. To turn this back into a reference
to the right decl's implicit OpaqueTypeDecl, use type reconstruction. Since type reconstruction
doesn't normally concern itself with non-type decls, set up a lookup table in SourceFiles and
ModuleFiles to let us handle the mapping from mangled name to opaque type decl in type
reconstruction.

(Since we're invoking type reconstruction during type checking, when the module hasn't yet been
fully validated, we need to plumb a LazyResolver into the ASTBuilder in an unsightly way. Maybe
there's a better way to do this... Longer term, at least, this surface design gives space for
doing things more the right way--a more request-ified decl validator ought to be able to naturally
lazily service this request without the LazyResolver reference, and if type reconstruction in
the future learns how to reconstruct non-type decls, then the lookup tables can go away.)

jckarter committed:
merge fixup

jckarter committed:
Relax some assertions to allow opaque-type values at global scope in scripts.

rdar://problem/49818962

omochi commented:
Please tell me how to build this branch?
I want to trace these new functions with debugger(Xcode) to learn compiler internals. 

(I am interested in generated resilient representation of opaque types and
these implementation of generation processes.)

But I failed several times so ask here now.
In particular, what is correct options which is passed to `utils/update-checkout`.
(In my experience, some feature branches are not compatible with latest swift-llvm, swift-clang if they experienced important update since feature development started. I recently face trouble around `clang::vfs` and `llvm::vfs` updates.)

I tried to sync with around starting date of these commits.

```
$ utils/update-checkout --tag swift-DEVELOPMENT-SNAPSHOT-2018-11-30-a
$ git checkout jckarter/opaque-type-runtime
$ utils/build-script --xcode --debug --skip-build-benchmarks --jobs 4
```

I got

```
Ld /Users/omochi/work/swift-source-ORT/build/Xcode-DebugAssert/swift-macosx-x86_64/Debug/lib/swift/macosx/x86_64/libswiftPhotos.dylib normal x86_64
clang: error: no such file or directory: '/Users/omochi/work/swift-source-ORT/build/Xcode-DebugAssert/swift-macosx-x86_64/Debug/lib/swift/macosx/x86_64/libswiftAVFoundation.dylib'
clang: error: no such file or directory: '/Users/omochi/work/swift-source-ORT/build/Xcode-DebugAssert/swift-macosx-x86_64/Debug/lib/swift/macosx/x86_64/libswiftCoreLocation.dylib'
clang: error: no such file or directory: '/Users/omochi/work/swift-source-ORT/build/Xcode-DebugAssert/swift-macosx-x86_64/Debug/lib/swift/macosx/x86_64/libswiftCoreMedia.dylib'
...
```

my environment
OS: macOS Mojave
xcode-select: Xcode-beta 10.2 beta 4

jckarter commented:
The branch is based off of master branch, so you should be able to `./utils/update-checkout --scheme master`, and then checkout this branch for Swift, and build.

jckarter committed:
Serialization: Handle reentrance while deserializing OpaqueTypeDecl. rdar://problem/49829836

jckarter commented:
Note though that there isn't any optimizer work in this branch, so you'll always be seeing the worst case performance where the underlying type is opaque to callers for the time being.

omochi commented:
Thank you very much!
I succeeded to build and run code using ORT.
I need one more step to checkout latest source of `Dispatch/Queue.swift`.

Finally:

```
$ utils/update-checkout --scheme master
$ git checkout jckarter/opaque-type-runtime
$ git checkout master -- stdlib/public/Darwin/Dispatch/Queue.swift
$ utils/build-script --debug
$ utils/build-script --debug --xcode
```

Ninja and Xcode are both ok.

OS: macOS Mojave 10.14.4
Xcode: Xcode 10.2 (10E125)

Thanks to tell me about status around performance.
In my understanding, current implementation generates always resilient archetype since this commit https://github.com/apple/swift/pull/22072/commits/0de8a394226329dc55ace75a03aabb6d2fa97827 .
Some tricks like specialization to use underlying type directly will be added in future.

jckarter committed:
Allow OpaqueTypeDecls to be (de)mangled as contexts, rdar://problem/49831658

jckarter commented:
@swift-ci Please build toolchain

jckarter commented:
@omochi That is correct.

jckarter committed:
merge fixup

jckarter commented:
@swift-ci Please build toolchain

jckarter commented:
@swift-ci Please build toolchain

jckarter committed:
Validate opaque return types for properties and subscripts.

jckarter committed:
Check the underlying type of get-only computed properties with opaque return types.

jckarter committed:
Serialize opaque return type of var/subscript decls.

jckarter committed:
Visit var/subscript opaque return decls during IRGen/TBDGen.

jckarter committed:
Track var/subscript opaque return decls for type reconstruction.

jckarter committed:
Demangler: Register substitution for opaque archetype manglings

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test Linux
